### ISSUE TITLE
Add ability to cancel execution

### ISSUE THREAD

**Issue by nedstoyanov** [App-vNext/Polly#46]
It may be handy to have ability to cancel execution for example, via a `CancelationToken`

**Comment by reisenberger** [comment]
@nedstoyanov In #56 I have another issue that might move things towards cancellation support (tho it's quite a recast of the Polly retry architecture).

We also think cancellation support would be a valuable addition. Exponential back-off retry (eg after 1, 2, 4 seconds; or even 5, 10, 30) is wonderful, especially in a machine-to-machine/service-to-service environment, where eventual consistency is acceptable, and eventual success (even after quite generous delays) is far preferable to faults to be manually troubleshot later.  However, the lack of cancellation support makes it hard to conduct a graceful shutdown of such retries.

**Comment by reisenberger** [comment]
@joelhulen Some starting thoughts / a possible spec around cancellation support. Keen to hear others' views.

[1] All types of asynchronous retry policy should support cancellation.  In respect of retries, cancellation should prevent further retries commencing, if the delegate being `ExecuteAsync()`d throws.  For `WaitAndRetryAsync()`, cancellation should also interrupt the `Task.Delay`.  All retry policies should also honour cancelled-before-first-try (I believe this is .NET framework approach - don't even start a `Task` if it's cancellation token is found cancelled, on entry)

[2] Should synchronous retry policies support a `CancellationToken`?  `CancellationToken` exists for .NET4, so cancellation could conceivably be added for .NET4 versions of _synchronous_ retry policies.  However, would lead to divergence in the synchronous API between .NET3.5 and .NET4+. Perhaps conceptually simpler (including for understanding of the product) to limit/align cancellation support to async?

[3] Specifying that a retry policy should observe a cancellation token, probably forms part of the retry overload signature?, eg

`var cancellableRetryPolicy = Policy.Handle<WhateverException>.Retry(3, myCancellationToken)`

[4] `action`s being `ExecuteAsync()`d obviously often support a `CancellationToken` too.  But there's possibly no need for Polly to get involved with that (eg with extra `ExecuteAsync()` overloads), can leave cancellation tokens to be captured by implicit closure, to reduce Polly interface surface, eg:

```
CancellationToken cancellationToken = ... 
var retryPolicy = Policy.Handle<Whatever>.WaitAndRetryAsync(new[] {
                   1.Seconds(),
                   2.Seconds(),
                   4.Seconds()
                }, cancellationToken);
var result = retryPolicy.ExecuteAsync(() => httpClient.GetAsync(url, cancellationToken));
```
 
I sketched an implementation roughly along these lines yesterday while commuting, but the meat of the work implementing (and tests) depends on what of #53 or #56 is merged first.  Also v happy if anyone with more time than me, or App-vNext, want to do this ...

**Comment by nedstoyanov** [comment]
@reisenberger I like this approach. I am happy to have a crack at this, haven't contributes to OSS before but have been playing around with `async` a fair bit.

**Comment by reisenberger** [comment]
Thanks @nedstoyanov.  Apologies for delay in replying: AppvNext have asked me to come onboard the project team and help respond to issues, pull requests, and forward-plan, and we've been reviewing the various requests (I’m co-ordinated v closely with @joelhulen and the rest of the AppvNext team on all this).

At the moment the plan is to get #53 merged (thanks @yevhen), then come to this, cancellation support.  @nedstoyanov , I would suggest hold off at the moment on any work on this, as I have second thoughts/doubts now about parts [3] and [4] of the above proposal, and may post on this in the coming days.

**Community shout-out**: If anyone has specific Use Cases to describe for cancellation support or ways they'd like to see it work, please comment in this issue.

**Comment by reisenberger** [comment]
@nedstoyanov Just to let you know that I have taken on the coding of this (async cancellation support) (I became aware of a number of potentials 'gotcha's while thinking it through, explored thru coding, and now half done).  However, it would be great to have your review / input / comment when I post the Pull Request!  (eg if meets your needs, is as you envisaged) (or indeed contribution on other issues)

Intend to get to this early in the new year.

Thanks.

**Comment by nedstoyanov** [comment]
no worries @reisenberger, I am happy to have a look once you are done. I think it will be a useful feature.

**Comment by reisenberger** [comment]
I have now committed an implementation of this to a branch reisenberger:AsyncCancellationSupport3, diff viewable as: 
https://github.com/App-vNext/Polly/compare/master...reisenberger:AsyncCancellationSupport3
 
If anyone gets the chance to review this ( @nedstoyanov from your original interest? ) ( @yevhen from your heavy previous involvement in the async code? ) , any review input gratefully received.

@joelhulen I'll pull request this in a few days pending any comments.

.

Compared to my 4 Dec proposal above, the final implementation adopts the following decisions:

[1] as before
[2] cancellation support for PCL and .NET 4.5 async only
[3] specifying an execution which honours cancellation now forms part of the call to `ExecuteAsync(...)`, not policy configuration.  This aligns with the recent `continueOnCapturedContext` implementation, and more cleanly separates policy configuration concerns from action execution concerns.  You can configure an async policy of whatever type.  Separately, you can run individual actions through that policy either with cancellation support, continueOnCapturedContext, or not, according to the needs of the delegate to execute.
[4] Because supplying the 'cancellationToken' is now part of `ExecuteAsync(...)`, the safer explicit capture (rather than implicit closure) was preferred to support delegates which in turn can honour a cancellationToken.

Example syntax:

```csharp
// Try several times to retrieve from a uri, but support cancellation at any time.
CancellationToken cancellationToken = // ...
var policy = Policy.Handle<Exception>().WaitAndRetryAsync(new[] { TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(5), TimeSpan.FromSeconds(10) });
var response = await policy.ExecuteAsync(ct => httpClient.GetAsync(uri, ct), cancellationToken);
```

**Comment by joelhulen** [comment]
Everything looks good to me. I say ship it if no one objects.

**Comment by nedstoyanov** [comment]
Had a brief look and it looks good. Will try and play with it a bit more later on. Good work.

**Comment by joelhulen** [comment]
@nedstoyanov: Just to make sure the loop is closed, I wanted to let you know that @reisenberger's implementation here has been merged into what is now version 3.0. Thanks for your review and great work you've contributed!
